server/src/engine/websocket.rs

```diff
@@ -0,0 +1,147 @@
+use actix_web_actors::ws;
+use serde::{Deserialize};
+use log::{debug, info, warn};
+use uuid::Uuid;
+
+use crate::game::{Player, player_message};
+use crate::engine::{Server, server_message};
+
+/// How often heartbeat pings are sent
+const HEARTBEAT_INTERVAL: Duration = Duration::from_secs(5);
+/// How long before lack of client response causes a timeout
+const CLIENT_TIMEOUT: Duration = Duration::from_secs(10);
+
+#[derive(Deserialize, Debug)]
+#[serde(tag = "action")]
+enum Message {
+    Reconnect { session_id: Uuid },
+    Connect { username: String },
+    GlobalChat { message: String },
+}
+
+pub struct WebSocket {
+    /// Client must send ping at least once per 10 seconds (CLIENT_TIMEOUT),
+    /// otherwise we drop connection.
+    hb: Instant,
+
+    server_addr: Addr<Server>,
+    pub player_addr: Option<Addr<Player>>,
+}
+
+impl Actor for WebSocket {
+    type Context = ws::WebsocketContext<Self>;
+
+    /// Method is called on actor start. We start the heartbeat process here.
+    fn started(&mut self, ctx: &mut Self::Context) {
+        self.hb(ctx);
+    }
+
+    fn stopped(&mut self, _ctx: &mut Self::Context) {
+        if let Some(player_addr) = &self.player_addr {
+            player_addr.do_send(player_message::Disconnected {});
+        }
+    }
+}
+
+/// Handler for `ws::Message`
+impl StreamHandler<Result<ws::Message, ws::ProtocolError>> for WebSocket {
+    fn handle(
+        &mut self,
+        msg: Result<ws::Message, ws::ProtocolError>,
+        ctx: &mut Self::Context,
+    ) {
+        match msg {
+            Ok(ws::Message::Ping(msg)) => {
+                self.hb = Instant::now();
+                ctx.pong(&msg);
+            }
+            Ok(ws::Message::Pong(_)) => {
+                self.hb = Instant::now();
+            }
+            Ok(ws::Message::Text(text)) => {
+                match serde_json::from_str(text.as_str()) {
+                    Ok(msg) => self.handle_message(ctx, msg),
+                    Err(_) => warn!("Unprocessable input '{}'", text)
+                }
+            }
+            Ok(ws::Message::Binary(_bin)) => {
+                warn!("Received binary data")
+            }
+            Ok(ws::Message::Close(_)) => {
+                ctx.stop();
+            }
+            _ => ctx.stop(),
+        }
+    }
+}
+
+impl WebSocket {
+    pub fn new(server_addr: Addr<Server>) -> Self {
+        Self {
+            hb: Instant::now(),
+            server_addr,
+            player_addr: None,
+        }
+    }
+
+    /// helper method that sends ping to client every second.
+    /// also this method checks heartbeats from client
+    fn hb(&self, ctx: &mut <Self as Actor>::Context) {
+        ctx.run_interval(HEARTBEAT_INTERVAL, |act, ctx| {
+            // check client heartbeats
+            if Instant::now().duration_since(act.hb) > CLIENT_TIMEOUT {
+                info!("Websocket Client heartbeat failed, disconnecting!");
```

Comment:
Wasn't sure either. The heartbeat is quite strict right now and drops the connection after 10 seconds. Since the client code is prepared to handle disconnects I considered it info.

Reply:
:man_shrugging: ...I think it's fine for now, you can always change it later